### PR TITLE
Add `catch-error-name` rule

### LINKED ISSUE #10: Enforce error name in `catch`
I want to enforce the use of `err` as the error name in try/catch and promises:

```js
try {} catch (err) {}
```

```js
foo.catch(err => {});
```

-

Should be easy to make it fixable.

### PR DESCRIPTION
**Pull request by jamestalmage**
Fixes #10.

Disregards if there are nested CatchClauses.

### CODE REVIEW

File: rules/caught-error-name.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+		'CatchClause:exit': function (node) {
+			if (!stack.pop()) {
+				context.report({
+					node: node.param,
+					message: 'The catch clause parameter should be named ' + name
+				});
+			}
+		}
+	};
+};
```
Comment (sindresorhus) [inline rules/caught-error-name.js]: Can you provide a options schema to validate the option: https://github.com/sindresorhus/eslint-plugin-xo/blob/1c287ab3e0a0c9a57a4194eac6b17d8bbeb7fda6/rules/filename-case.js#L49-L60

File: rules/caught-error-name.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+	var stack = [];
+
+	return {
+		'CatchClause': function (node) {
+			if (stack.length === 1) {
+				stack[0] = true;
+			}
+			stack.push(stack.length > 0 || node.param.name === name);
+		},
+		'CatchClause:exit': function (node) {
```
Comment (sindresorhus) [inline rules/caught-error-name.js]: Why does it need to wait until exiting the node?

File: rules/caught-error-name.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+			if (!stack.pop()) {
+				context.report({
+					node: node.param,
+					message: 'The catch clause parameter should be named ' + name
```
Comment (sindresorhus) [inline rules/caught-error-name.js]: End it with a `.`

File: rules/caught-error-name.js (hunk 2 above)
Comment (sindresorhus) [inline rules/caught-error-name.js]: To ignore nested catch clauses, right?

File: rules/caught-error-name.js (hunk 3 above)
Comment (sindresorhus) [inline rules/caught-error-name.js]: Include autofix?

Should be as easy as:

```js
fix: function (fixer) {
	return fixer.replaceText(node.param, name);
}
```

**Comment by sindresorhus** [comment]
Can it support Promise catch too? `foo.catch(err => {});`

**Comment by sindresorhus** [comment]
`catch-error-name`?

File: rules/caught-error-name.js (hunk 3 above)
Comment (sindresorhus) [inline rules/caught-error-name.js]: http://eslint.org/docs/developer-guide/working-with-rules#applying-fixes

**Comment by sindresorhus** [comment]
Needs some docs, but I'd be happy to fix that if you want.

File: rules/caught-error-name.js (hunk 3 above)
Comment (jamestalmage) [inline rules/caught-error-name.js]: You would also need to rename all other references to it, which isn't so easy with ESLints current fixer implementation.

File: rules/caught-error-name.js (hunk 2 above)
Comment (jamestalmage) [inline rules/caught-error-name.js]: Yes

File: rules/caught-error-name.js (hunk 3 above)
Comment (sindresorhus) [inline rules/caught-error-name.js]: Oh, that's a good point. I guess we can skip it for now then.

**Comment by jfmengels** [comment]
LGTM, except for the points already mentionned by @sindresorhus :)

**Comment by jamestalmage** [comment]
> Can it support Promise catch too? `foo.catch(err => {});`

Without type information? Not with 100% reliability. But I think it's pretty safe to assume `.catch(FunctionExpression)` means it's a Promise.

**Comment by jamestalmage** [comment]
Updated.
Now handles promises. 
Added documentation.

**Comment by jfmengels** [comment]
LGTM :+1:

**Comment by sindresorhus** [comment]
LGTM2. Yay :)

68747470733a2f2f6d65646961342e67697068792e636f6d2f6d656469612f75344163337942656b623272362f67697068792e676966

**Pull request merged**